File management and UI fixes

stevecl5 opened the pull request:
This branch fixes a bug that prevented Workbench from running local simulations (Issue #83). It also resolves an issue where CSS resources could not be found if the Workbench JAR was moved to a different directory.

Additionally, it improves the functionality of the Simulation Specification dialog:
- better input validation for hostname and simulation folder
- improved messaging for invalid input
- Run and Test Connection buttons are now only enabled when valid input is present

stevecl5 commented:
Added:
- fixes issue #15
- cleanup of variable names and event handlers

stiber commented on WorkbenchProject/src/main/java/edu/uwb/braingrid/workbench/model/SimulationSpecification.java:
```diff
@@ -128,30 +128,30 @@ public String getSimExecutable() {
-     * downloaded to, and where it is built.
+     * Provides the location of the folder where the simulator is downloaded, built, and executed.
+     * If the simulator execution is specified to take place on a remote machine, this is relative
+     * to the root directory for the user.
```
For clarification, I think you mean "home directory" for the user. There is only one root: "/".

stevecl5 replied:
Good catch. I copied that description from the existing @return tag and didn't notice the error. I corrected this in both places in my latest commit.

stiber commented on WorkbenchProject/src/main/java/edu/uwb/braingrid/workbench/model/SimulationSpecification.java:
```diff
@@ -240,20 +241,20 @@ public String getUsername() {
-     * set in the specification) cloned/checked out to, built, and executed.
+     * Sets the location of the folder where the simulator is downloaded, built, and executed. If
+     * the simulator execution is specified to take place on a remote machine, this is relative to
+     * the root directory for the user.
```
Ditto from above.

stiber commented on WorkbenchProject/src/main/java/edu/uwb/braingrid/workbench/ui/ScriptSpecificationDialog.java:
```diff
@@ -509,10 +507,10 @@ public ScriptSpecificationDialog(boolean modal, SimulationSpecification simSpec)
         String userName = simSpec.getUsername();
         String folder = simSpec.getSimulatorFolder();
         String codeLocation = simSpec.getCodeLocation();
+        codeLocation = codeLocation == null || codeLocation.isEmpty()
+                ? DEFAULT_REPO_URI : codeLocation;
```
OK, two comments:

1. It's probably best to use some parentheses on complex expressions, rather than rely on operator precedence, for the sake of code readability. For example, I'd write `(codeLocation == null || codeLocation.isEmpty())`.
2. Given that the "else" part of the trinary assignment leaves `codeLocation` unchanged, is this really simpler than:
```C++
if (codeLocation == null || codeLocation.isEmpty()) codeLocation = DEFAULT_REPO_URI;
```

stevecl5 replied:
To add some context, my only change to this line was replacing `getDefaultCodeLocation()` with `DEFAULT_REPO_URI` (since `getDefaultCodeLocation()` was a private helper function that did nothing but return a constant). However, I agree that using parentheses for the complex expression is a good practice here.

A single-line if statement does not fit the style guide, so the alternative would be:
```java
if (codeLocation == null || codeLocation.isEmpty()) {
    codeLocation = DEFAULT_REPO_URI;
}
```

I think using the if statement version is a slight improvement, but either way is fine with me.

stiber commented:
This looks good to me overall. I admit I don't have a deep command of the code these days. I can approve, or @king-shak can concur; just let me know.

stevecl5 commented:
@stiber
I know that @king-shak has already done some testing on this branch, and we will be going over it together later today. If we are both happy with everything, I would like to do the merge on Wednesday after our weekly meeting so that the main branch is up to date with the latest fixes and features.

stevecl5 commented:
Added:
- updated all* occurrences of "state output" and "history dump" to "result file" for consistency with Graphitti
- updated result file name from prepending `historyDump_` to appending `-out` for consistency with Graphitti

\* The XML tag in the simulation config file is still named `stateOutputFileName` for now since this is required in order to run BrainGrid. At some point, this will need to be changed to `resultFileName` in order to work with Graphitti.

king-shak reviewed (approved):
All looks good! Tested the changes and everything works as expected.
